### ISSUE TITLE
MethodChannel to mock ResultType.noAppToOpen in test

### ISSUE THREAD

**Issue by jadensiow93** [crazecoder/open_file#180]
Hello,

Currently I am trying to mock the ResultType.noAppToOpen in the test as shown below. 
![image](https://user-images.githubusercontent.com/87740572/152756655-ab7fdb5e-06ff-42f9-9d84-610a7042a2eb.png)

So I did the method channel below.
`const MethodChannel('open_file').setMockMethodCallHandler(methodCall) { 
return OpenResult(message: 'No APP found to open this file.' , type: ResultType.noAppToOpen);`

But I am unable to have the desired return of `ResultType.noAppToOpen` in my test. 
Am I doing the methodChannel wrongly for this since there is filePath inside `OpenFile.open(filePath)` ? Or is the methodChannel not working for this. Thank you.

**Comment by ValentinVignal** [comment]
The issue is the you are not running your test on iOS or Android.
So the condition 
https://github.com/crazecoder/open_file/blob/2cadda6d282935afce8b02aefe41a198f0c49c0c/lib/src/plaform/open_file.dart#L23

is `true`.

If you were running on iOS and Android, mocking the channel `open_file` would work because it is calling 
https://github.com/crazecoder/open_file/blob/2cadda6d282935afce8b02aefe41a198f0c49c0c/lib/src/plaform/open_file.dart#L57

at the end.

The issue here is that there is no way no mock whatever is inside the `if` of 

https://github.com/crazecoder/open_file/blob/2cadda6d282935afce8b02aefe41a198f0c49c0c/lib/src/plaform/open_file.dart#L23

It would be nice if this package could include some test examples.
I created an issue for that : https://github.com/crazecoder/open_file/issues/181

**Comment by burhanrashid52** [comment]
I think having an API where the target platform is checked by the `BuildContext` will help to replace the variant in testing using `variant:` param in `testWidget`. More [here](https://medium.com/geekculture/using-platform-variants-when-testing-flutter-widgets-d3a750d452f0)

Something like this. `OpenFile.open(context,filePath);`

**Comment by burhanrashid52** [comment]
@ValentinVignal This is also a fair solution. But I don't think it's backward compatible and might break a lot of code if used extensively in the app.
Need something like this might work.
```dart
static Future<OpenResult> open(String filePath,{BuildContext context}) {
   final platform = context?.targetPlatform ?? getPlatformFromStaticCheck();
}  
```

**Comment by burhanrashid52** [comment]
> @ValentinVignal This is also a fair solution. But I don't think it's backward compatible and might break a lot of code if used extensively in the app. Need something like this might work.
> 
> ```dart
> static Future<OpenResult> open(String filePath,{BuildContext context}) {
>    final platform = context?.targetPlatform ?? getPlatformFromStaticCheck();
> }  
> ```

If @crazecoder is happy with the above solution then I can create a PR for it.

**Comment by ValentinVignal** [comment]
@burhanrashid52 Yes my suggestion is a breaking change, but if you increase the version number to `4.0.0` and mention it in the change logs it should be fine right ?

The issue your approach is that we would have to carry this `context` a bit everywhere, even in service classes/methods that are not supposed to use the `BuildContext` (not UI).

**Comment by burhanrashid52** [comment]
> @burhanrashid52 Yes my suggestion is a breaking change, but if you increase the version number to `4.0.0` and mention it in the change logs it should be fine right ?

I will leave that up to the author of this library. However, I personally don't prefer `OpenFile.instance` pattern. Seems a lot verbose to me.

> The issue your approach is that we would have to carry this context a bit everywhere, even in service classes/methods that are not supposed to use the BuildContext (not UI).

In the example, I share, `BuildContext` is optional. We can abstract out the platform check to a separate class, and based on parameters we can decide which implementation to use and use that abstract class throughout the library

**Comment by GP4cK** [comment]
I agree with https://github.com/crazecoder/open_file/issues/180#issuecomment-1044603096, I'd rather not have to pass around the context which is UI related in my services. If it comes at the cost of having the code a little more verbose, I'm willing to pay the price.

**Comment by burhanrashid52** [comment]
Hmm... Maybe using a static instance inside the static `open` method might solve both issues.

**Comment by ValentinVignal** [comment]
> Hmm... Maybe using a static instance inside the static `open` method might solve both issues.

It would yes. But for it to be mocked, it would need to use a public method from the instance

**Comment by crazecoder** [comment]
I think it was fixed in 3.5.0, close it now, and I will reopen it  if there are any problems
